lib/firenest/synced_server.ex

```diff
@@ -0,0 +1,446 @@
+  end
+
+  ## Implementation
+
+  @impl true
+  def init({mod, arg, topology, name}) do
+    state = %{
+      name: name,
+      topology: topology,
+      mod: mod,
+      int: nil,
+      awaiting_hello: [],
+      awaiting_up: [],
+      replicas: []
+    }
+
+    with {:ok, state} <- init_mod(mod, arg, state),
+         {:ok, state, node_ref} <- sync_named(topology, name, state) do
+      Process.put(__MODULE__, {topology, name, node_ref})
+      {:ok, state}
+    end
+  end
+
+  @impl true
+  def handle_cast(msg, state) do
+    {:stop, {:badcast, msg}, state}
+  end
+
+  @impl true
+  def handle_call(msg, from, state) do
+    case apply_callback(state, :handle_call, [msg, from]) do
+      {:reply, reply, int} -> {:reply, reply, %{state | int: int}}
+      {:reply, reply, int, timeout} -> {:reply, reply, %{state | int: int}, timeout}
+      {:stop, reason, reply, int} -> {:stop, reason, reply, %{state | int: int}}
+      other -> handle_common(other, state)
+    end
+  end
+
+  @impl true
+  def handle_info({:named_up, node_ref, name}, %{name: name} = state) do
+    %{awaiting_up: awaiting, replicas: replicas} = state
+
+    case delete_element(awaiting, node_ref) do
+      {:ok, awaiting} ->
+        result = apply_callback(state, :handle_replica, [:up, node_ref])
+        handle_common(result, %{state | awaiting_up: awaiting, replicas: [node_ref | replicas]})
+
+      :error ->
+        {:noreply, %{state | awaiting_hello: [node_ref | awaiting]}}
+    end
+  end
+
+  def handle_info({:named_down, node_ref, name}, %{name: name} = state) do
+    %{awaiting_hello: awaiting, replicas: replicas} = state
+
+    case delete_element(replicas, node_ref) do
+      {:ok, replicas} ->
+        result = apply_callback(state, :handle_replica, [:down, node_ref])
+        handle_common(result, %{state | replicas: replicas})
+
+      :error ->
+        {:noreply, %{state | awaiting_hello: List.delete(awaiting, node_ref)}}
+    end
+  end
+
+  def handle_info({__MODULE__, :hello, node_ref}, state) do
+    %{awaiting_hello: awaiting, replicas: replicas} = state
+
+    case delete_element(awaiting, node_ref) do
+      {:ok, awaiting} ->
+        result = apply_callback(state, :handle_replica, [:up, node_ref])
+
+        handle_common(result, %{state | awaiting_hello: awaiting, replicas: [node_ref | replicas]})
+
+      :error ->
+        {:noreply, %{state | awaiting_up: [node_ref | awaiting]}}
+    end
+  end
+
+  def handle_info({__MODULE__, :message, msg, node_ref}, %{replicas: replicas} = state) do
+    if node_ref in replicas do
+      result = apply_callback(state, :handle_remote, [msg, node_ref])
+      handle_common(result, state)
+    else
+      {:noreply, state}
+    end
+  end
+
+  def handle_info(msg, state) do
+    result = apply_callback(state, :handle_info, [msg])
+    handle_common(result, state)
+  end
+
+  @impl true
+  def terminate(reason, state) do
+    apply_optional_callback(state, :terminate, [reason])
```

Comment:
This function is only called once. Maybe we should inline its implementation here like we do for `code_change/3`? Anyway, this is completely up to you. :)